[RFC] Fix oldtest on Windows

erw7 opened the pull request:
This is an attempt to fix oldtest on Windows.

erw7 committed:
Change to set shellslash to test under same conditions as vim

erw7 committed:
Remove code that is no longer needed by set shellslash

erw7 committed:
Fix get_path_cutoff() on Windows

Fix an issue where the result of get_path cutoff() was incorrect when
using set shellslash.

erw7 committed:
Add target fixff to testdir/Makefile

In Windows, since conversion of the line feed code is necessary, porting
fixff target from Make_mingw.mak of vim.

erw7 committed:
Change test execution conditions

Since neovim needs iconv to process cp932, it changes the test execution
condition.

erw7 committed:
Remove test52

Test52 does nothing on UNIX. And vim on Windows performs mbyte-conversion
between Unicode and code page using Windows API(See :h mbyte-conversion).
Test52 seems to be testing its functionality. In neovim, the conversion
function by Windows API has been removed. Therefore, there is no point
in doing this test.

erw7 commented:
The test52 will succeed in the next patch, but I think it is better to remove it.

```diff
--- a/src/nvim/testdir/test52.in
+++ b/src/nvim/testdir/test52.in
@@ -7,27 +7,27 @@ STARTTEST
 :" write tests:
 :" combine three values for 'encoding' with three values for 'fileencoding'
 :" also write files for read tests
+:e ++enc=utf-8
 /^1
-:set encoding=utf-8
 :.w! ++enc=utf-8 test.out
 :.w ++enc=cp1251 >>test.out
 :.w ++enc=cp866 >>test.out
 :.w! ++enc=utf-8 Xutf8
 /^2
-:set encoding=cp1251
+:e ++enc=cp1251
 :.w ++enc=utf-8 >>test.out
 :.w ++enc=cp1251 >>test.out
 :.w ++enc=cp866 >>test.out
 :.w! ++enc=cp1251 Xcp1251
 /^3
-:set encoding=cp866
+:e ++enc=cp866
 :.w ++enc=utf-8 >>test.out
 :.w ++enc=cp1251 >>test.out
 :.w ++enc=cp866 >>test.out
 :.w! ++enc=cp866 Xcp866
 :"
 :" read three 'fileencoding's with utf-8 'encoding'
-:set encoding=utf-8 fencs=utf-8,cp1251
+:set fencs=utf-8,cp1251
 :e Xutf8
 :.w ++enc=utf-8 >>test.out
 :e Xcp1251
@@ -37,7 +37,7 @@ STARTTEST
 :.w ++enc=utf-8 >>test.out
 :"
 :" read three 'fileencoding's with cp1251 'encoding'
-:set encoding=utf-8 fencs=utf-8,cp1251
+:set fencs=utf-8,cp1251
 :e Xutf8
 :.w ++enc=cp1251 >>test.out
 :e Xcp1251
@@ -47,7 +47,7 @@ STARTTEST
 :.w ++enc=cp1251 >>test.out
 :"
 :" read three 'fileencoding's with cp866 'encoding'
-:set encoding=cp866 fencs=utf-8,cp1251
+:set fencs=utf-8,cp1251
 :e Xutf8
 :.w ++enc=cp866 >>test.out
 :e Xcp1251
```

justinmk commented:
> The test52 will succeed in the next patch, but I think it is better to remove it.

👍  That test was missed when we removed `'encoding'` because it only runs on Windows.

> When using `set binary`, change to use `++bin` argument because it is not
`nobinary` when returning to the original file with `:e #` on line 18 of
test42.in.

Maybe `set cpo+=S` is needed? https://github.com/neovim/neovim/issues/10730#issuecomment-519885658

## Status

Waiting on https://github.com/neovim/neovim/pull/10650. cc @blueyed

erw7 committed:
Change value of cpo [skip travis]

set cpo+=S was added because the test failed because the cpo value
was different from vim(See #10730).

erw7 commented:
> Maybe set cpo+=S is needed?

After all, `set cpo+=S` was necessary. With #10662 and this pull request, all of oldtest succeeded on Windows :tada:.

justinmk commented:
> should I set it in `setup.vim`?

Seems like a good idea. I wonder why Vim's tests are running with a different 'cpo'? 

> With #10662 and this pull request, all of oldtest succeeded on Windows 🎉.

Beautiful! Thanks for figuring this out.

erw7 commented:
> I wonder why Vim's tests are running with a different 'cpo'?

The following description exists in Vim: starting.txt.

> Using the "-u" argument with another argument than DEFAULTS has the side effect that the 'compatible' option will be on by default.  This can have unexpected effects.

Tests are run like `vim -u dos.vim` or `vim -u unix.vim`. Also, `set nocompatible` is not done in any of `dos.vim`, `unix.vim` and `setup.vim`. From the above, vim tests are run with `compatible`, therefore, the `cpo` during test is different between neovim and vim.

erw7 commented:
>> should I set it in setup.vim?

> Seems like a good idea.

So after #10650 is merged, change it that way.

erw7 commented:
#10764 ? After that, some new Windows oldtests seem to fail. Is there a missing patch?

#10764 
https://ci.appveyor.com/project/neovim/neovim/builds/26676309/job/g7kgwc3kjx45s9ie#L16117

before
https://ci.appveyor.com/project/neovim/neovim/builds/26675459/job/1an3i4gd6alwsxrc#L16075

erw7 committed:
Fix test failure on Windows [skip travis]

justinmk commented on src/nvim/testdir/shared.vim:
```diff
@@ -202,9 +202,17 @@ endfunc
+      return '..\\..\\..\\build\\bin\\nvim.exe'
+    else
+      return '../../../build/bin/nvim'
+    endif
```
Let's just use `v:progpath` here. Vim's reason for not using it is weak.

erw7 committed:
Change to use v:progpath instead of constant [skip ci]
